File: typescript/src/data-structures/disjoint-set.ts
/**
 * Disjoint Set (Union-Find) implementation with path compression and union by rank.
 * Time: O(α(n)) amortized for all operations (nearly constant)
 * Space: O(n)
 */

export class DisjointSet {
  private parent: number[];
  private rank: number[];
  private _count: number;

  constructor(size: number) {
    this.parent = new Array(size);
    this.rank = new Array(size).fill(0);
    this._count = size;

    for (let i = 0; i < size; i++) {
      this.parent[i] = i;
    }
  }

  find(x: number): number {
    if (this.parent[x] !== x) {
      this.parent[x] = this.find(this.parent[x]); // Path compression
    }
    return this.parent[x];
  }

  union(x: number, y: number): boolean {
    const rootX = this.find(x);
    const rootY = this.find(y);

    if (rootX === rootY) {
      return false;
    }

    // Union by rank
    if (this.rank[rootX] < this.rank[rootY]) {
      this.parent[rootX] = rootY;
    } else if (this.rank[rootX] > this.rank[rootY]) {
      this.parent[rootY] = rootX;
    } else {
      this.parent[rootY] = rootX;
      this.rank[rootX]++;
    }

    this._count--;
    return true;
  }

  connected(x: number, y: number): boolean {
    return this.find(x) === this.find(y);
  }

  count(): number {
    return this._count;
  }

  size(): number {
    return this.parent.length;
  }
}

File: swift/Sources/DSA/DataStructures/DisjointSet.swift
/// Disjoint Set (Union-Find) implementation with path compression and union by rank.
/// Time: O(α(n)) amortized for all operations (nearly constant)
/// Space: O(n)
public struct DisjointSet {
    private var parent: [Int]
    private var rank: [Int]
    private var setCount: Int

    public init(_ size: Int) {
        parent = Array(0..<size)
        rank = Array(repeating: 0, count: size)
        setCount = size
    }

    public mutating func find(_ x: Int) -> Int {
        if parent[x] != x {
            parent[x] = find(parent[x]) // Path compression
        }
        return parent[x]
    }

    public mutating func union(_ x: Int, _ y: Int) -> Bool {
        let rootX = find(x)
        let rootY = find(y)

        if rootX == rootY { return false }

        // Union by rank
        if rank[rootX] < rank[rootY] {
            parent[rootX] = rootY
        } else if rank[rootX] > rank[rootY] {
            parent[rootY] = rootX
        } else {
            parent[rootY] = rootX
            rank[rootX] += 1
        }

        setCount -= 1
        return true
    }

    public mutating func connected(_ x: Int, _ y: Int) -> Bool {
        find(x) == find(y)
    }

    public var count: Int {
        setCount
    }

    public var size: Int {
        parent.count
    }
}
